storage/tree/node_id2.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tree
+
+import "fmt"
+
+// NodeID2 is a faster NodeID that does zero memory allocations.
+//
+// TODO(pavelkalinnikov): Rename to NodeID and document it properly when the
+// code has migrated.
+type NodeID2 struct {
+	path string
+	last byte
+	bits byte
+}
+
+// NewNodeID2 creates a NodeID2 from the given path bytes truncated to the
+// specified number of bits if necessary. Panics if the number of bits is
+// negative, or is more than the byte string contains.
+func NewNodeID2(path string, bits int) NodeID2 {
+	if bits == 0 {
+		return NodeID2{}
+	} else if mx := len(path) * 8; bits > mx {
+		panic(fmt.Sprintf("NewNodeID2: bits %d > %d", bits, mx))
+	}
+	bytes, tail, mask := split(bits)
+	last := path[bytes] & mask
+	return NodeID2{path: path[:bytes], last: last, bits: byte(tail)}
```

Comment:
It *may* be worth a comment here on the fact that go creates substrings without the need for further allocation because the underling arrays are immutable.